Add handling `export =` for namespaces

dts-bundle-generator@1.1.0

to reproduce

```bash
git clone https://github.com/e2tox/reproduce-dts-bundle-generator-node-imports.git
cd reproduce-dts-bundle-generator-node-imports
npm install
npm test
```

Error details:
```bash
> dts-bundle-generator -o ./index.d.ts ./stoppable.d.ts

Writing generated file to ./index.d.ts...
Checking the generated file...
index.d.ts(1,52): TS2304: Cannot find name 'EventEmitter'.
Error: Compiled with errors
npm ERR! Test failed.  See above for more details.

```

--- a/src/bundle-generator.ts
+++ b/src/bundle-generator.ts
@@ -310,16 +310,26 @@ function shouldNodeBeImported(node: ts.NamedDeclaration, rootFileExports: Readon
 
 	// we should import only symbols which are used in types directly
 	return Array.from(symbolsUsingNode).some((symbol: ts.Symbol) => {
-		if (symbol.valueDeclaration === undefined && symbol.declarations === undefined) {
-			return false;
-		} else if (symbol.valueDeclaration !== undefined && isDeclarationFromExternalModule(symbol.valueDeclaration)) {
-			return false;
-		} else if (symbol.declarations !== undefined && symbol.declarations.every(isDeclarationFromExternalModule)) {
+		const symbolsDeclarations = getDeclarationsForSymbol(symbol);
+		if (symbolsDeclarations.length === 0 || symbolsDeclarations.every(isDeclarationFromExternalModule)) {
 			return false;
 		}
 
 		return rootFileExports.some((rootSymbol: ts.Symbol) => typesUsageEvaluator.isSymbolUsedBySymbol(symbol, rootSymbol));
 	});
+}
+
+function getDeclarationsForSymbol(symbol: ts.Symbol): ts.Declaration[] {
+	const result: ts.Declaration[] = [];
+	if (symbol.valueDeclaration !== undefined) {
+		result.push(symbol.valueDeclaration);
+	}
+
+	if (symbol.declarations !== undefined) {
+		result.push(...symbol.declarations);
+	}
+
+	return result;
 }
 
 function isDeclarationFromExternalModule(node: ts.Declaration): boolean {

--- a/src/bundle-generator.ts
+++ b/src/bundle-generator.ts
@@ -248,16 +248,17 @@ function addImport(statement: ts.DeclarationStatement, library: string, imports:
 		throw new Error(`Import/usage unnamed declaration: ${statement.getText()}`);
 	}
 
-	const importName = statement.name.getText();
-	normalLog(`Adding import with name "${importName}" for library "${library}"`);
-
 	let libraryImports = imports.get(library);
 	if (libraryImports === undefined) {
 		libraryImports = new Set<string>();
 		imports.set(library, libraryImports);
 	}
 
-	libraryImports.add(importName);
+	const importName = statement.name.getText();
+	if (!libraryImports.has(importName)) {
+		normalLog(`Adding import with name "${importName}" for library "${library}"`);
+		libraryImports.add(importName);
+	}
 }
 
 function getRootSourceFile(program: ts.Program): ts.SourceFile {

--- a/src/bundle-generator.ts
+++ b/src/bundle-generator.ts
@@ -8,6 +8,7 @@ import {
 	hasNodeModifier,
 	isDeclareGlobalStatement,
 	isDeclareModuleStatement,
+	isNamespaceStatement,
 	isNodeNamedDeclaration,
 } from './typescript-helpers';
 
@@ -90,6 +91,15 @@ export function generateDtsBundle(filePath: string, options: GenerationOptions =
 				isStatementUsed: (statement: ts.Statement) => isNodeUsed(statement, rootFileExports, typesUsageEvaluator, typeChecker),
 				shouldStatementBeImported: (statement: ts.DeclarationStatement) => shouldNodeBeImported(statement, rootFileExports, typesUsageEvaluator),
 				getModuleInfo: (fileName: string) => getModuleInfo(fileName, criteria),
+				getDeclarationsForExportedAssignment: (exportAssignment: ts.ExportAssignment) => {
+					const symbolForExpression = typeChecker.getSymbolAtLocation(exportAssignment.expression);
+					if (symbolForExpression === undefined) {
+						return [];
+					}
+
+					const symbol = getActualSymbol(symbolForExpression, typeChecker);
+					return getDeclarationsForSymbol(symbol);
+				},
 			},
 			collectionResult
 		);
@@ -153,8 +163,10 @@ interface UpdateParams {
 	isStatementUsed(statement: ts.Statement): boolean;
 	shouldStatementBeImported(statement: ts.DeclarationStatement): boolean;
 	getModuleInfo(fileName: string): ModuleInfo;
+	getDeclarationsForExportedAssignment(exportAssignment: ts.ExportAssignment): ts.Declaration[];
 }
 
+// tslint:disable-next-line:cyclomatic-complexity
 function updateResult(params: UpdateParams, result: CollectingResult): void {
 	if (params.currentModule.type === ModuleType.ShouldNotBeUsed) {
 		return;
@@ -180,6 +192,11 @@ function updateResult(params: UpdateParams, result: CollectingResult): void {
 			continue;
 		}
 
+		if (ts.isExportAssignment(statement) && statement.isExportEquals && params.currentModule.type === ModuleType.ShouldBeImported) {
+			updateResultForImportedEqExportAssignment(statement, params, result);
+			continue;
+		}
+
 		if (!params.isStatementUsed(statement)) {
 			verboseLog(`Skip file member: ${statement.getText().replace(/(\n|\r)/g, '').slice(0, 50)}...`);
 			continue;
@@ -195,6 +212,29 @@ function updateResult(params: UpdateParams, result: CollectingResult): void {
 		} else if (params.currentModule.type === ModuleType.ShouldBeInlined) {
 			result.statements.push(statement);
 		}
+	}
+}
+
+function updateResultForImportedEqExportAssignment(exportAssignment: ts.ExportAssignment, params: UpdateParams, result: CollectingResult): void {
+	const moduleDeclarations = params.getDeclarationsForExportedAssignment(exportAssignment)
+		.filter(isNamespaceStatement)
+		.filter((s: ts.ModuleDeclaration) => s.getSourceFile() === exportAssignment.getSourceFile());
+
+	// if we have `export =` somewhere so we can decide that every declaration of exported symbol in this way
+	// is "part of the exported module" and we need to update result according every member of each declaration
+	// but treat they as current module (we do not need to update module info)
+	for (const moduleDeclaration of moduleDeclarations) {
+		if (moduleDeclaration.body === undefined || !ts.isModuleBlock(moduleDeclaration.body)) {
+			continue;
+		}
+
+		updateResult(
+			{
+				...params,
+				statements: moduleDeclaration.body.statements,
+			},
+			result
+		);
 	}
 }
 

--- a/src/typescript-helpers.ts
+++ b/src/typescript-helpers.ts
@@ -41,3 +41,10 @@ export function isDeclareModuleStatement(statement: ts.Statement): statement is 
 export function isDeclareGlobalStatement(statement: ts.Statement): statement is ts.ModuleDeclaration {
 	return ts.isModuleDeclaration(statement) && Boolean(statement.flags & ts.NodeFlags.GlobalAugmentation);
 }
+
+/**
+ * Returns whether node is `namespace` ModuleDeclaration
+ */
+export function isNamespaceStatement(node: ts.Node): node is ts.ModuleDeclaration {
+	return ts.isModuleDeclaration(node) && Boolean(node.flags & ts.NodeFlags.Namespace);
+}

--- a/tests/test-cases/handle-export-eq-from-npm/config.ts
+++ b/tests/test-cases/handle-export-eq-from-npm/config.ts
@@ -1,0 +1,5 @@
+import { TestCaseConfig } from '../../test-cases/test-case-config';
+
+const config: TestCaseConfig = {};
+
+export = config;

--- a/tests/test-cases/handle-export-eq-from-npm/input.ts
+++ b/tests/test-cases/handle-export-eq-from-npm/input.ts
@@ -1,0 +1,11 @@
+import { SomeCoolInterface } from 'package-with-export-eq';
+import { NamedDeclaration } from 'typescript';
+import { EventEmitter } from 'events';
+
+export class StoppableEventEmitter extends EventEmitter {
+	public emitStoppableEvent(): this {
+		return this;
+	}
+}
+
+export type ExportType = SomeCoolInterface | NamedDeclaration | string;

--- a/tests/test-cases/handle-export-eq-from-npm/output.d.ts
+++ b/tests/test-cases/handle-export-eq-from-npm/output.d.ts
@@ -1,0 +1,8 @@
+import { EventEmitter } from 'events';
+import { SomeCoolInterface } from 'package-with-export-eq';
+import { NamedDeclaration } from 'typescript';
+
+export declare class StoppableEventEmitter extends EventEmitter {
+	emitStoppableEvent(): this;
+}
+export declare type ExportType = SomeCoolInterface | NamedDeclaration | string;

